src/modules/kvs/kvs.c

```diff
@@ -1223,6 +1229,68 @@ static int heartbeat_root_cb (struct kvsroot *root, void *arg)
     return 0;
 }
 
+static void sync_continuation (flux_future_t *f, void *arg)
+{
+    struct kvs_ctx *ctx = arg;
+    struct kvsroot *root;
+
+    if (flux_rpc_get (f, NULL) < 0)
+        flux_log_error (ctx->h, "%s: flux_rpc_get", __FUNCTION__);
+    flux_future_destroy (ctx->sync_f);
+    ctx->sync_f = NULL;
+
+    /* on rank 0, primary namespace has to be here */
+    root = kvsroot_mgr_lookup_root (ctx->krm, KVS_PRIMARY_NAMESPACE);
+    assert (root);
+
+    /* if there are no commits to be processed, shut down timer, we'll
+     * wake it up next time the primary KVS namespace has work to be
+     * done on it */
+    if (kvstxn_mgr_ready_transaction_count (root->ktm) == 0) {
+        flux_watcher_stop (ctx->sync_w);
+        ctx->sync_timer_running = false;
+    }
+}
+
+static void sync_cb (flux_reactor_t *r,
+                     flux_watcher_t *w,
+                     int revents,
+                     void *arg)
+{
+    struct kvs_ctx *ctx = arg;
+    flux_kvs_txn_t *txn = NULL;
+    flux_future_t *f = NULL;
+
+    /* prior sync not completed, skip this turn in the sync */
+    if (ctx->sync_f)
+        return;
+
+    /* N.B. no actual changes in this transaction */
+    if (!(txn = flux_kvs_txn_create ())) {
+        flux_log_error (ctx->h, "%s: flux_kvs_txn_create", __FUNCTION__);
```

Comment:
The log messages in this function are all equally unlikely and don't provide much info.  Maybe all the failure cases could use one better message like "periodic checkpoint setup failed: ..."